lib/Dialect/RTG/Transforms/ElaborationPass.cpp

```diff
@@ -0,0 +1,568 @@
+
+  /// Cache the builders to continue insertions at their current insertion point
+  /// for the reason stated above.
+  DenseMap<Block *, OpBuilder> builderPerBlock;
+};
+
+/// Used to signal to the elaboration driver whether the operation should be
+/// removed.
+enum class DeletionKind { Keep, Delete };
+
+/// Interprets the IR to perform and lower the represented randomizations.
+class Elaborator : public RTGOpVisitor<Elaborator, FailureOr<DeletionKind>,
+                                       function_ref<void(Operation *)>> {
+public:
+  using RTGBase = RTGOpVisitor<Elaborator, FailureOr<DeletionKind>,
+                               function_ref<void(Operation *)>>;
+  using RTGBase::visitOp;
+
+  Elaborator(SymbolTable &table, std::mt19937 &rng) : rng(rng) {}
+
+  /// Helper to perform internalization and keep track of interpreted value for
+  /// the given SSA value.
+  template <typename ValueTy, typename... Args>
+  void internalizeResult(Value val, Args &&...args) {
+    // TODO: this isn't the most efficient way to internalize
+    auto ptr = std::make_unique<ValueTy>(std::forward<Args>(args)...);
+    auto *e = ptr.get();
+    auto [iter, _] = interned.insert({e, std::move(ptr)});
+    state[val] = iter->second.get();
+  }
+
+  /// Print a nice error message for operations we don't support yet.
+  FailureOr<DeletionKind>
+  visitUnhandledOp(Operation *op,
+                   function_ref<void(Operation *)> addToWorklist) {
+    return op->emitError("elaboration not supported");
+  }
+
+  FailureOr<DeletionKind>
+  visitExternalOp(Operation *op,
+                  function_ref<void(Operation *)> addToWorklist) {
+    // TODO: we only have this to be able to write tests for this pass without
+    // having to add support for more operations for now, so it should be
+    // removed once it is not necessary anymore for writing tests
+    if (op->use_empty()) {
+      for (auto &operand : op->getOpOperands()) {
+        Value val = materializer.materialize(state.at(operand.get()),
+                                             op->getBlock(), op->getLoc());
+        if (!val)
+          return op->emitError("failed to materialize value");
+        operand.set(val);
+      }
+      return DeletionKind::Keep;
+    }
+
+    return visitUnhandledOp(op, addToWorklist);
+  }
+
+  FailureOr<DeletionKind>
+  visitOp(SetCreateOp op, function_ref<void(Operation *)> addToWorklist) {
+    SetVector<ElaboratorValue *> set;
+    for (auto val : op.getElements())
+      set.insert(state.at(val));
+
+    internalizeResult<SetValue>(op.getSet(), std::move(set),
+                                op.getSet().getType());
+    return DeletionKind::Delete;
+  }
+
+  FailureOr<DeletionKind>
+  visitOp(SetSelectRandomOp op, function_ref<void(Operation *)> addToWorklist) {
+    auto *set = cast<SetValue>(state.at(op.getSet()));
+
+    size_t selected;
+    std::uniform_int_distribution<size_t> dist(0, set->getSet().size() - 1);
```

Comment:
This distribution is apparently not very well specified and may behave differently for different compilers and libraries. So it seems like I need to run my own distribution...

Reply:
Seems like it was a great idea to change to `rtg.elaboration_custom_seed` otherwise I probably wouldn't have noticed that for a long time.